Improving Ground Temperature Models

mitchute opened the pull request:
This adds three new ground temperature model objects. It also pull the rest of the ground temperature objects in so the same methods are used.

A few noteworthy points for reviewers:
-    I create and remove a new environment [here](https://github.com/mmatts/EnergyPlus/blob/1579840-GroundTemps/src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc#L128) which I use to read annual weather data from the weather file. I didn't see any issues with it on my machine, but someone more familiar with manipulating environment level variables might see that it is overkill, or not enough.
-    Related to the previous point: I also had to put a few flags [here](https://github.com/mmatts/EnergyPlus/blob/1579840-GroundTemps/src/EnergyPlus/WeatherManager.cc#L1717) and [here](https://github.com/mmatts/EnergyPlus/blob/1579840-GroundTemps/src/EnergyPlus/WeatherManager.cc#L4977) so when I called `ManageWeather()` weather data would not be reported. This was causing issues with the eio file.
-    Use of `shared_ptr`. Another chance for review on this subject. 

To-do:
- [ ] Address reviewer comments
- [ ] Add transition source

Commit:
Initial commit

Commit:
Merge branch 'develop' of github.com:NREL/EnergyPlus into 1579840-GroundTemps

Commit:
Base factory methods

Commit:
Adding KusudaGroundTempsModel

Commit:
Merge branch 'develop' of github.com:NREL/EnergyPlus into 1579840-GroundTemps

Commit:
Transitioning slinkyGLHE undisturbed ground temps to new KA model

Commit:
Transitioning inputs to accept ground temperature models for PipingSystem:Underground:Domain

Commit:
Switching PipingSystem:Underground:Domain and PipingSystem_Underground_FHX.idf to use the new ground temperature models. Some small diffs introduced that need to be addressed.

Commit:
Resolving diffs in PipingSystem_Underground_FXH.idf. This will cause some minor diff in the slinky model due the fact that previously the undisturbed ground temp was only being calculated using the DayOfSim variable, instead of the actual sim time down to the timestep level. Some cleanup as well.

Commit:
Preliminary work updating slab to use undisturbed ground temperature models

Commit:
Merge branch 'develop' of github.com:NREL/EnergyPlus into 1579840-GroundTemps

Commit:
Updating input files for PipingSystem:Underground:Domain, Site:GroundDomain:Slab, and Site:GroundDomain:Basement objects. Also fixed small error in slab/basement code. No diffs in slab and piping system files. Basement IDD still needs to be updated.

Commit:
Updating inputs for Site:GroundDomain:Basement for accepting ground temperature models

Commit:
Updating input for GroundHeatExchanger:HorizontalTrench to use new ground temperature models

Commit:
Stub-outs for ground temperature model unit tests

rbojja committed:
Adding first unit test.

Commit:
Cleaning up and moving code allowing for Site:GroundTemperature:Shallow to be used when KA correlation was left blank. Still some errors that need to be resolved.

Commit:
Merge branch '1579840-GroundTemps' of github.com:mmatts/EnergyPlus into 1579840-GroundTemps

Commit:
Removing unused variables.
Cleaning up the getInput for Site:GroundDomain:Shallow objects

Commit:
Fixes get input error for basements.

Commit:
Moving calls to GetGroundTempModelAndInit after all calls to current GetObjectItem is through. This was causing errors.

Commit:
Updating Pipe:Underground to accept different ground temperature models

Commit:
First pass at updating documentation.

Commit:
Stubbing out for moving Site:GroundTemperature:* objects from the WeatherManager to the GroundTempsManager

Commit:
Moving code for Site:GroundTemperature:Shallow from WeatherManager.cc to the GTM code.

Commit:
Updating to allow months and seconds to be passed as inputs. Changed so diffusivity is passed on construction.

Commit:
Moving code for Site:GroundTemperature:BuildingSurfaces from the WeatherManager to the new GTM code

Commit:
Moving code for Site:GroundTemperature:FCFactorMethod from WeatherManager to new GTM code

Commit:
First pass at moving Site:GroundTemperature:Deep to GTM code. Still has one 'vtable' reference error that needs to be cleaned up.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Commit:
Adding code that was missed for Site:GroundTemperatur:Deep

Commit:
Initial work on the FD model.

Commit:
More work filling in structure of FD model.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Conflicts:
	doc/src/docs/InputOutputReference/01a-InputOutputReference.md

Commit:
First pass at creating new environment to be used for reading in the weather data.

Commit:
Implemented FD method code. Validated against analytical solution. Now on to surface and lower domain boundary conditions.

Commit:
Reads weather data from file. Simulation runs, and results look reasonable. Need additional validation.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Conflicts:
	doc/src/docs/InputOutputReference/01b-InputOutputReference.md
	src/EnergyPlus/PlantPipingSystemsManager.cc

Commit:
Added Xing 5 parameter 2 harmonic model. Cleaned up surface boundary for FD model.

Commit:
Finished adding Xing model 5 parameter 2 harmonic model. Added comment blocks. Updated getGroundTemp for FiniteDiffGroundTempsModel to interpolate to correct data.

Commit:
Beginning working on unit tests

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Commit:
Intermittent commit so I can revert back in case I break it after restructuring.

Commit:
Moving files into separate directory

Commit:
Moving files around. Breaking out into 1 class per file.

Commit:
Removed old files.

Commit:
Cleaning up factory methods and GetGroundTempModelAndInit signatures.

Commit:
More work on unit tests.

Commit:
More unit testing work.

Commit:
Updating docs. Code cleanup. Adding unit tests.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Commit:
Adding example files. Finalizing and restructuring unit tests. Minor doc tweaks.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Commit:
Removing unintended diffs

Commit:
Cleaning up warnings and failures.
Summary: removed unused variables--need to remove some others but need to discuss with reviewers first; FCFactorMethod ground temps were not being read from EPW header--this is corrected now; environment was not being reset properly after running FD GTM--corrected now.

Commit:
More cleanup.
Fixed one issue with the FC Factor Method ground temps where they were being put on the eio when they were not needed.
Fixed the Slab object in the UnitarySystem_WaterCoils_wMultiSpeedFan.idf file.

Myoldmopar commented:
@mmatts When you need a 10 second distraction, could you update the title of this PR so it is easier for users to understand?

mitchute commented:
@Myoldmopar better?

Myoldmopar commented:
:+1:

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Commit:
Putting auto & reference back where needed.

Commit:
Cleaning up string diffs caused by uppercase object names.

Myoldmopar commented:
@mmatts So maybe you are already looking at this, but I took a minute and found why those test files were failing.  Looks like right [here](https://github.com/mmatts/EnergyPlus/blob/1579840-GroundTemps/src/EnergyPlus/WeatherManager.cc#L7158) you are accessing the errorsFound property of the siteFCFactorMethodGroundTempsPtr variable.  But, it looks like in the init routine, you can return a [nullptr](https://github.com/mmatts/EnergyPlus/blob/1579840-GroundTemps/src/EnergyPlus/GroundTemperatureModeling/GroundTemperatureModelManager.cc#L107).  This is fine to indicate there wasn't any of those, but you'll want to check whether it is null before trying to access any members of that null pointer:

```
if ( siteFCFactorMethodGroundTempsPtr ) {
  ErrorsFound = siteFCFactorMethodGroundTempsPtr->errorsFound;
}
```

Anyway, you may already be on this, just pointing it out.

Commit:
Adding protection for null pointers.

Myoldmopar commented:
@mmatts Those files are still crashing.  Your pointer protection is correct in the location you put it, but there are other spots like [this](https://github.com/mmatts/EnergyPlus/blob/82337bc6e5cd9d77307660f5968714c82add93e8/src/EnergyPlus/WeatherManager.cc#L2053-L2059) where you still access the null pointer if those entries weren't in the input.  Search the code for those variables and add protection everywhere you try to access them.

mbadams5 commented on src/EnergyPlus/GroundTemperatureModeling/GroundTemperatureModelManager.hh:
```diff
@@ -0,0 +1,64 @@
+
+	extern Array1D_string const CurrentModuleObjects;
+
+	static std::vector < std::shared_ptr < BaseGroundTempsModel > > groundTempModels;
```
@mmatts This might not be doing what you expect or want (maybe it is, I don't know). If you have a static variable in a header then every time you include that header you get a new, unique version of that variable. Read [this](https://stackoverflow.com/questions/92546/variable-declarations-in-header-files-static-or-not) for more information.

@DeadParrot @Myoldmopar

Myoldmopar commented:
Hmm, yeah, I don't think it needs to be static at all.  It's just in the namespace.  I'd suggest taking out the static.  (@mmatts @mbadams5)

mbadams5 commented on tst/EnergyPlus/unit/Fixtures/GroundTempsFixture.hh:
```diff
@@ -0,0 +1,39 @@
+		}
+
+		virtual void TearDown() {
+			GroundTemperatureManager::clear_state();		
```
@mmatts I would lean towards removing this fixture and adding this clear_state() in EnergyPlusFixture instead. Then use that fixture in your tests. Also you only need to include GroundTemperatureModelManager.hh in the fixture and then in the unit test file you can include the other header files.

mbadams5 replied:
Also, do each of the ground temperature methods need their own unit test file? You could put them in a single GroundTemperatureModels.unit.cc or something. But that is just a minor quibble. I don't really have strong feelings about that.

Myoldmopar replied:
_(The separated files was my suggestion to do one class per file and then mirror that in the unit tests.)_

mbadams5 replied:
@Myoldmopar that is fine. Like I said, I don't have strong feelings about it.

mbadams5 commented on src/EnergyPlus/GroundTemperatureModeling/KusudaAchenbachGroundTemperatureModel.cc:
```diff
@@ -0,0 +1,262 @@
+					Real64 minSurfTemp( 100 ); // Set high month 1 temp will be lower and actually get updated
+					Real64 maxSurfTemp( -100 ); // Set low initially but will get updated
+
+					std::shared_ptr< BaseGroundTempsModel > shallowObj = GetGroundTempModelAndInit( CurrentModuleObjects( objectType_SiteShallowGroundTemp ), "" );
```
@mmatts does this need to be a shared_ptr?

mitchute replied:
@mbadams5 It is possible that other objects use Site:GroundTemperature:Shallow. Only one is allowed in the input file so this (or the other objects) would be using the same Site:GroundTemperature:Shallow object, which seemed like an appropriate use of shared_ptr. Is that correct?

mbadams5 replied:
@mmatts I understand in general there is only one Site:GroundTemperature:Shallow object per IDF, but in this specific instance you are creating a new object and using it below and no where else. It isn't getting copied anywhere, only used to generate currMonthTemp, thus you really only need unique_ptr. But your factory method returns a shared_ptr ... hummm... let me see if there is a way around that. I think there is.

mbadams5 replied:
Ok I remembered (and did a little research). If you make the factory return unique_ptr instead of shared_ptr, you can 

``` cpp
// converts std::unique_ptr to std::shared_ptr
std::shared_ptr< BaseGroundTempsModel > sp =
GetGroundTempModelAndInit( CurrentModuleObjects( objectType_SiteShallowGroundTemp ), "" );
```

But you can also then use the same factory method to generate unique_ptrs. 

That said, I now understand and see what you are doing. You are returning a shared_ptr from a vector< shared_ptr >, if already created. Hummm ok. The way you are doing it requires a shared_ptr so I think this is fine.

Slightly on topic segue, another option is to return std::weak_ptr instead of shared_ptr. This will not increase the reference count of the shared_ptr and is allowed to dangle. Read [here](http://en.cppreference.com/w/cpp/memory/weak_ptr) and [here](https://stackoverflow.com/questions/4984381/shared-ptr-and-weak-ptr-differences) for more information. I don't think we should do this for this PR, but is food for thought moving forward.

DeadParrot replied:
And this is how `shared_ptr` spreads like a virus. Because an object is _used_ by multiple other objects is not justification for `shared_ptr`: only if they must share ownership because the object must stay alive until the last of them is done with it (or itself destroyed) do you need/want shared ownership. I didn't look into this instance so I'm not sure the use case for the factory is returning pre-existing objects but that is unusual: normally I'd think of such behavior as occurring outside the factory and only calling the factory if you determine you "need a new one". But this situation may justify the unusual design.

Normally the pattern of a factory returning `unique_ptr` is best: it can be loaded into a `shared_ptr` be the callee as @mbadams5 shows.

mbadams5 commented on src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.hh:
```diff
@@ -0,0 +1,199 @@
+			Real64 simTimeInDays;
+
+		// Default constructor
+		FiniteDiffGroundTempsModel() :
```
@mmatts Since you are using a factory method to create these objects, this should probably be private instead of public. This is the same comment for all of your derived classes. Also does all the data in the class need to be public or can some of it be private? Are any of the member variables common among all the derived classes? If so, they could move to the base class and be 'protected' visibility.

mbadams5 commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -281,6 +282,11 @@ namespace WeatherManager {
 	Array1D< SpecialDayData > SpecialDays;
 	Array1D< DataPeriodData > DataPeriods;
 
+	std::shared_ptr< BaseGroundTempsModel > siteShallowGroundTempsPtr;
```
@mmatts Are these used anywhere outside of this file? Do they need to be shared_ptrs? It doesn't seem like they need to be.

Myoldmopar replied:
Waiting to chime in until @mmatts gives a response...

mitchute replied:
@Myoldmopar @mbadams5 the shallow object is used, like you saw above by one of the other ground temp models for certain situations. @Myoldmopar feel free to chime in if you feel inclined on this. I will be out for a bit so I will respond when I get back.

Myoldmopar replied:
So the factory method should create the new shared pointer in the ground temp manager, and store it in the vector.  Then anything else that later needs a ground temp model can call the factory method and share ownership of that model instance, and call methods to get ground temperature.  I think that is the proper place for a shared_ptr, right?

mbadams5 replied:
@Myoldmopar Yeah I see what he is doing now. You are right, it needs to be shared_ptr for this use case.

mbadams5 commented:
@mmatts @Myoldmopar @DeadParrot @kbenne 

This looks pretty good. I do have questions about the need for shared_ptrs though. From looking at the code it seems like you needed the shared_ptrs for certain things. But what if you refactored the code so it avoided the need for those shared_ptrs? Can you refactor the code to avoid them? (rhetorical question) These are more general questions about object ownership and lifetime, which need to be answered sooner than later.

Myoldmopar commented:
@mbadams5 

Thanks for all these comments.  We do need answers on that.  We are planning to have a more formal chat _after_ the release to:
- Describe the OO stuff starting to form up inside E+
- Evaluate pros/cons, propose modifications and plans
- Fix up the existing approaches and plan out propagation to the rest of E+

I'd just like to remind that we are on feature freeze "_now_".  To me, 
- that does _not_ mean we let _bad_ code into develop, but 
- it _does_ mean we don't rigor too hard over things that work properly but could use a different design.

kthxbye

Commit:
Removed GroundTempsFixture.hh. Put clear_state on EnergyPlusFixture.

Commit:
Adding defaults for Site:GroundTemperature:FCFactorMethod. Small warnings cleanup.

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Conflicts:
	doc/src/docs/InputOutputReference/01b-InputOutputReference.md
	tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc

Myoldmopar commented:
Hmmm, OK @mmatts you were right that it would throw eio diffs for the missing FC entries.  But seeing those zeros added in there isn't good.  I think we'll back to backtrack and instead of assigning default values, just allow it to be null if there aren't FC values in the input file.  Or maybe better yet, just don't write the zero values to the eio file.  Then the variable can live in the program with a valid address, you just won't throw diffs.  I actually like that better.

mitchute commented:
@Myoldmopar I had the same thought when I saw it. I will push that change now.

Myoldmopar commented:
Yeah, I thought about it walking in today.  That should match the previous behavior exactly.  The values would have stayed at zero on initialization, but they never would've been reported.  OK cool.

Commit:
Cleanup eio reporting for unused FCFactorMethod ground temps.

Myoldmopar commented:
Also, did you address @mbadams5 comment about the static variable in the namespace?  I think ultimately it was just a matter of removing the static qualifier.  For this use case, since it isn't in a class, you don't need the static.

mitchute commented:
Yes, I removed the static, removed the GroundTempFixture, and put the clear_state() and other unit tests on the EnergyPlusFixture, [here](https://github.com/mmatts/EnergyPlus/commit/9ade595fc6a424f9cdd481ab54829ce5c419e67e).

mbadams5 commented:
@mmatts I don't see the removed static in that commit.

Make sure you change from static to extern and then put the variable in GroundTemperatureModelManager.cc file. Like how the other variables are declared and used in the header.

mitchute commented:
@mbadams5 good catch. I know I changed that... not sure why it didn't get saved. It will be pushed up shortly.

Commit:
Remove static from groundTempModels vector.

Myoldmopar commented:
@mmatts Here is what I am getting so far for the slab object.  

```
Site:GroundDomain:Slab,
    CoupledSlab, !- Name
    5, !- Ground Domain Depth {m}
    1, !- Aspect Ratio
    5, !- Domain Perimeter Offset {m}
    1.8, !- Soil Thermal Conductivity {W/m-K}
    3200, !- Soil Density {kg/m3}
    836, !- Soil Specific Heat {J/kg-K}
    30, !- Soil Moisture Content Volume Fraction {percent}
    50, !- Soil Moisture Content Volume Fraction at Saturation {percent}
    15.5, !- Kusuda-Achenbach Average Surface Temperature {C}
    12.8, !- Kusuda-Achenbach Average Amplitude of Surface Temperature {C}
    17.3, !- Kusuda-Achenbach Phase Shift of Minimum Surface Temperature {days}
    1, !- Evapotranspiration Ground Cover Parameter
    GroundCoupledOSCM, !- Name of Floor Boundary Condition Model
    OnGrade, !- Slab Location (InGrade/OnGrade)
    , !- Slab Material Name
    , !- Horizontal Insulation (Yes/No)
    , !- Horizontal Insulation Material Name
    , !- Full Horizontal or Perimeter Only (Full/Perimeter)
    , !- Perimeter insulation width (m)
    Yes, !- Vertical Insulation (Yes/No)
    Insulation, !- Vertical Insulation Name
    1.5, !- Vertical perimeter insulation depth from surface (m)
    Hourly; !- Domain Simulation Interval. (Timestep/Hourly)
```

to 

```
  Site:GroundDomain:Slab,
    CoupledSlab,             !- Name
    5,                       !- Ground Domain Depth {m}
    1,                       !- Aspect Ratio
    5,                       !- Perimeter Offset {m}
    1.8,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    836,                     !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    1,                       !- Evapotranspiration Ground Cover Parameter
    GroundCoupledOSCM,       !- Slab Boundary Condition Model Name
    OnGrade,                 !- Slab Location
    ,                        !- Slab Material Name
    ,                        !- Horizontal Insulation
    ,                        !- Horizontal Insulation Material Name
    ,                        !- Horizontal Insulation Extents
    ,                        !- Perimeter Insulation Width {m}
    Yes,                     !- Vertical Insulation
    Insulation,              !- Vertical Insulation Material Name
    1.5,                     !- Vertical Insulation Depth {m}
    Hourly;                  !- Simulation Timestep

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    1.8,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    836,                     !- Soil Specific Heat {J/kg-K}
    15.5,                    !- Average Soil Surface Temperature {C}
    12.8,                    !- Average Amplitude of Surface Temperature {deltaC}
    17.3;                    !- Phase Shift of Minimum Surface Temperature {days}
```

Look OK?

I'm pushing on to the rest of the objects now.  You'll have to do some additional commenting on the new IDD fields to add \note, \default, \type and that sort of thing.  You might want to make a quick pass at it, and then @mjwitte will surely have additional recommendations.

mitchute commented:
Looks good. I will go update the idd fields.

Myoldmopar commented:
Basement:

```
  Site:GroundDomain:Basement,
    CoupledBasement,         !- Name
    5,                       !- Ground Domain Depth {m}
    1,                       !- Aspect ratio {m}
    1,                       !- Perimeter offset {m}
    1.8,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    836,                     !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    15.5, !- Kusuda-Achenbach Average Surface Temperature {C}
    12.8, !- Kusuda-Achenbach Average Amplitude of Surface Temperature {C}
    17.3, !- Kusuda-Achenbach Phase Shift of Minimum Surface Temperature {days}
    0.5,  !- Evapotranspiration Ground Cover Parameter
    BasementFloorOSCM,       !- Name of Basement Floor Boundary Condition Model
    Yes,                     !- Basement Horizontal Underfloor Insulation Present (Yes/No)
    Basement Insulation,     !- Basement Horizontal Insulation Underfloor Material Name
    Perimeter,               !- Full Horizontal or Perimeter Only (Full/Perimeter)
    1,                       !- Perimeter width (m)
    2.5,                     !- Depth of Basement Wall In Ground Domain {m}
    BasementWallOSCM,        !- Name of Basement Floor Boundary Condition Model
    Yes,                     !- Basement Wall Vertical Insulation Present(Yes/No)
    Basement Insulation,     !- Basement Wall Vertical Insulation Material Name
    2.3,                     !- Vertical insulation depth from surface (m)
    timestep;                !- Domain Update interval. (Timestep, Hourly);
```

to 

```
 Site:GroundDomain:Basement,
    CoupledBasement,         !- Name
    5,                       !- Ground Domain Depth {m}
    1,                       !- Aspect Ratio
    1,                       !- Perimeter Offset {m}
    1.8,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    836,                     !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    0.5,                     !- Evapotranspiration Ground Cover Parameter
    BasementFloorOSCM,       !- Basement Floor Boundary Condition Model Name
    Yes,                     !- Horizontal Insulation
    Basement Insulation,     !- Horizontal Insulation Material Name
    Perimeter,               !- Horizontal Insulation Extents
    1,                       !- Perimeter Horizontal Insulation Width {m}
    2.5,                     !- Basement Wall Depth {m}
    BasementWallOSCM,        !- Basement Wall Boundary Condition Model Name
    Yes,                     !- Vertical Insulation
    Basement Insulation,     !- Basement Wall Vertical Insulation Material Name
    2.3,                     !- Vertical Insulation Depth {m}
    timestep;                !- Simulation Timestep

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    1.8,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    836,                     !- Soil Specific Heat {J/kg-K}
    15.5,                    !- Average Soil Surface Temperature {C}
    12.8,                    !- Average Amplitude of Surface Temperature {deltaC}
    17.3;                    !- Phase Shift of Minimum Surface Temperature {days}
```

Myoldmopar commented:
And for @mjwitte's sake before he asks, the `KATemp 1` increments so no name conflict issues later.  As long as they don't have 100 slabs!!!

Myoldmopar commented:
PIpingSystem:Underground:Domain changes:

```
  PipingSystem:Underground:Domain,
    My Piping System,        !- Name
    4,                       !- Xmax {m}
    2.5,                     !- Ymax {m}
    75,                      !- Zmax {m}
    2,                       !- X-Direction Mesh Density Parameter
    Uniform,                 !- X-Direction Mesh Type
    ,                        !- X-Direction Geometric Coefficient
    2,                       !- Y-Direction Mesh Density Parameter
    Uniform,                 !- Y-Direction Mesh Type
    ,                        !- Y-Direction Geometric Coefficient
    6,                       !- Z-Direction Mesh Density Parameter
    Uniform,                 !- Z-Direction Mesh Type
    ,                        !- Z-Direction Geometric Coefficient
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    15.5,                    !- Kusuda-Achenbach Average Surface Temperature {C}
    12.8,                    !- Kusuda-Achenbach Average Amplitude of Surface Temperature {C}
    17.3,                    !- Kusuda-Achenbach Phase Shift of Minimum Surface Temperature {days}
    No,                      !- This Domain Includes Basement Surface Interaction
    ,                        !- Width of Basement Floor in Ground Domain {m}
    ,                        !- Depth of Basement Wall In Ground Domain {m}
    ,                        !- Shift Pipe X Coordinates By Basement Width
    ,                        !- Name of Basement Wall Boundary Condition Model
    ,                        !- Name of Basement Floor Boundary Condition Model
    0.005,                   !- Convergence Criterion for the Outer Cartesian Domain Iteration Loop {deltaC}
    100,                     !- Maximum Iterations in the Outer Cartesian Domain Iteration Loop
    0.408,                   !- Evapotranspiration Ground Cover Parameter
    1,                       !- Number of Pipe Circuits Entered for this Domain
    My Pipe Circuit;         !- Pipe Circuit 1
```

to

```
  PipingSystem:Underground:Domain,
    My Piping System,        !- Name
    4,                       !- Xmax {m}
    2.5,                     !- Ymax {m}
    75,                      !- Zmax {m}
    2,                       !- X-Direction Mesh Density Parameter
    Uniform,                 !- X-Direction Mesh Type
    ,                        !- X-Direction Geometric Coefficient
    2,                       !- Y-Direction Mesh Density Parameter
    Uniform,                 !- Y-Direction Mesh Type
    ,                        !- Y-Direction Geometric Coefficient
    6,                       !- Z-Direction Mesh Density Parameter
    Uniform,                 !- Z-Direction Mesh Type
    ,                        !- Z-Direction Geometric Coefficient
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    No,                      !- This Domain Includes Basement Surface Interaction
    ,                        !- Width of Basement Floor in Ground Domain {m}
    ,                        !- Depth of Basement Wall In Ground Domain {m}
    ,                        !- Shift Pipe X Coordinates By Basement Width
    ,                        !- Name of Basement Wall Boundary Condition Model
    ,                        !- Name of Basement Floor Boundary Condition Model
    0.005,                   !- Convergence Criterion for the Outer Cartesian Domain Iteration Loop {deltaC}
    100,                     !- Maximum Iterations in the Outer Cartesian Domain Iteration Loop
    0.408,                   !- Evapotranspiration Ground Cover Parameter
    1,                       !- Number of Pipe Circuits Entered for this Domain
    My Pipe Circuit;         !- Pipe Circuit 1

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    2,                       !- Soil Thermal Conductivity {W/m-K}
    Uniform,                 !- Soil Density {kg/m3}
    ,                        !- Soil Specific Heat {J/kg-K}
    ,                        !- Average Soil Surface Temperature {C}
    6,                       !- Average Amplitude of Surface Temperature {deltaC}
    Uniform;                 !- Phase Shift of Minimum Surface Temperature {days}
```

Moving on...

mitchute commented:
@Myoldmopar Hold up. That one has some issues.

Myoldmopar commented:
Dangit beat me to it!  Here's what it does now :smile:

```
  PipingSystem:Underground:Domain,
    My Piping System,        !- Name
    4,                       !- Xmax {m}
    2.5,                     !- Ymax {m}
    75,                      !- Zmax {m}
    2,                       !- X-Direction Mesh Density Parameter
    Uniform,                 !- X-Direction Mesh Type
    ,                        !- X-Direction Geometric Coefficient
    2,                       !- Y-Direction Mesh Density Parameter
    Uniform,                 !- Y-Direction Mesh Type
    ,                        !- Y-Direction Geometric Coefficient
    6,                       !- Z-Direction Mesh Density Parameter
    Uniform,                 !- Z-Direction Mesh Type
    ,                        !- Z-Direction Geometric Coefficient
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Volume Fraction {percent}
    50,                      !- Soil Moisture Content Volume Fraction at Saturation {percent}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    No,                      !- This Domain Includes Basement Surface Interaction
    ,                        !- Width of Basement Floor in Ground Domain {m}
    ,                        !- Depth of Basement Wall In Ground Domain {m}
    ,                        !- Shift Pipe X Coordinates By Basement Width
    ,                        !- Name of Basement Wall Boundary Condition Model
    ,                        !- Name of Basement Floor Boundary Condition Model
    0.005,                   !- Convergence Criterion for the Outer Cartesian Domain Iteration Loop {deltaC}
    100,                     !- Maximum Iterations in the Outer Cartesian Domain Iteration Loop
    0.408,                   !- Evapotranspiration Ground Cover Parameter
    1,                       !- Number of Pipe Circuits Entered for this Domain
    My Pipe Circuit;         !- Pipe Circuit 1

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    15.5,                    !- Average Soil Surface Temperature {C}
    12.8,                    !- Average Amplitude of Surface Temperature {deltaC}
    17.3;                    !- Phase Shift of Minimum Surface Temperature {days}
```

mitchute commented:
:+1:

Commit:
Updating idd fields for all objects using undisturbed ground temps.

Myoldmopar commented:
Whew, this was the worst yet.  For the Pipe:Underground object, I had to mine the soil property information out of the `Material` objects in the idf.  Here are the results:

```
  Material,
    Buried Pipe Soil,        !- Name
    Smooth,                  !- Roughness
    1.5,                     !- Thickness {m}
    0.36,                    !- Conductivity {W/m-K}
    2000.0,                  !- Density {kg/m3}
    1200.0,                  !- Specific Heat {J/kg-K}
    0.9,                     !- Thermal Absorptance
    0.5,                     !- Solar Absorptance
    0.5;                     !- Visible Absorptance

  Pipe:Underground,
    Pipe Heat Transfer Towers,  !- Name
    Insulated Buried Pipe,   !- Construction Name
    Condenser Tower Outlet Node,  !- Fluid Inlet Node Name
    HTPipe Outlet Node,      !- Fluid Outlet Node Name
    SunExposed,              !- Sun Exposure
    0.05,                    !- Pipe Inside Diameter {m}
    200.0,                   !- Pipe Length {m}
    Buried Pipe Soil,        !- Soil Material Name
    13,                      !- Average Soil Surface Temperature {C}
    1.5,                     !- Amplitude of Soil Surface Temperature {C}
    30;                      !- Phase Constant of Soil Surface Temperature {days}
```

to:

```
  Pipe:Underground,
    Pipe Heat Transfer Towers,  !- Name
    Insulated Buried Pipe,   !- Construction Name
    Condenser Tower Outlet Node,  !- Fluid Inlet Node Name
    HTPipe Outlet Node,      !- Fluid Outlet Node Name
    SunExposed,              !- Sun Exposure
    0.05,                    !- Pipe Inside Diameter {m}
    200.0,                   !- Pipe Length {m}
    Buried Pipe Soil,        !- Soil Material Name
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1;                !- Name of Undisturbed Ground Temperature Model

! Found a material to match the soil material; name =Buried Pipe Soil                                                                                    
  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    0.36,                    !- Soil Thermal Conductivity {W/m-K}
    2000.0,                  !- Soil Density {kg/m3}
    1200.0,                  !- Soil Specific Heat {J/kg-K}
    13,                      !- Average Soil Surface Temperature {C}
    1.5,                     !- Average Amplitude of Surface Temperature {deltaC}
    30;                      !- Phase Shift of Minimum Surface Temperature {days}
```

Myoldmopar commented:
Horizontal trench:

```
  GroundHeatExchanger:HorizontalTrench,
    My Pipe Circuit,         !- Name
    Plant Supply Intermediate Node,  !- Inlet Node Name
    Plant Supply Outlet Node,!- Outlet Node Name
    0.004,                   !- Design Flow Rate {m3/s}
    75,                      !- Trench Length in Pipe Axial Direction {m}
    2,                       !- Number of Trenches
    2.0,                     !- Horizontal Spacing Between Pipes {m}
    0.016,                   !- Pipe Inner Diameter {m}
    0.02667,                 !- Pipe Outer Diameter {m}
    1.25,                    !- Burial Depth {m}
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    0.3895,                  !- Pipe Thermal Conductivity {W/m-K}
    641,                     !- Pipe Density {kg/m3}
    2405,                    !- Pipe Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Percent {percent}
    50,                      !- Soil Moisture Content Percent at Saturation {percent}
    15.5,                    !- Kusuda-Achenbach Average Surface Temperature {C}
    12.8,                    !- Kusuda-Achenbach Average Amplitude of Surface Temperature {C}
    17.3,                    !- Kusuda-Achenbach Phase Shift of Minimum Surface Temperature {days}
    0.408;                   !- Evapotranspiration Ground Cover Parameter
```

to:

```
  GroundHeatExchanger:HorizontalTrench,
    My Pipe Circuit,         !- Name
    Plant Supply Intermediate Node,  !- Inlet Node Name
    Plant Supply Outlet Node,!- Outlet Node Name
    0.004,                   !- Design Flow Rate {m3/s}
    75,                      !- Trench Length in Pipe Axial Direction {m}
    2,                       !- Number of Trenches
    2.0,                     !- Horizontal Spacing Between Pipes {m}
    0.016,                   !- Pipe Inner Diameter {m}
    0.02667,                 !- Pipe Outer Diameter {m}
    1.25,                    !- Burial Depth {m}
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    0.3895,                  !- Pipe Thermal Conductivity {W/m-K}
    641,                     !- Pipe Density {kg/m3}
    2405,                    !- Pipe Specific Heat {J/kg-K}
    30,                      !- Soil Moisture Content Percent {percent}
    50,                      !- Soil Moisture Content Percent at Saturation {percent}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    0.408;                   !- Evapotranspiration Ground Cover Parameter

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    1.08,                    !- Soil Thermal Conductivity {W/m-K}
    962,                     !- Soil Density {kg/m3}
    2576,                    !- Soil Specific Heat {J/kg-K}
    15.5,                    !- Average Soil Surface Temperature {C}
    12.8,                    !- Average Amplitude of Surface Temperature {deltaC}
    17.3;                    !- Phase Shift of Minimum Surface Temperature {days}
```

Shikes.  One more to go, slinky:tm:.

Commit:
Reordered undisturbed ground temp objects. Other minor cleanup.

mitchute commented:
@Myoldmopar I added some idd comments here: https://github.com/mmatts/EnergyPlus/commit/1f44b41cf30351f09587b8c66184b5683aa7f7c2

Would you like addition information added?

Myoldmopar commented:
I'll rely on @mjwitte if it is about the idd; I tried rewriting him as @mjwitte-bot, but it didn't go so well.  

However, I am done.  Here is the slinky:

```
  GroundHeatExchanger:Slinky,
    Slinky GHX,       !- Name
    GHE Inlet Node,   !- Inlet Node
    GHE Outlet Node,  !- Outlet Node
    0.0033,           !- Design Flow Rate [m3/s]
    1.2,              !- Soil Thermal Conductivity [W/m-K]
    3200,             !- Soil Density [kg/m3]
    850,              !- Soil Specific Heat [J/kg-K]
    1.8,              !- Pipe Thermal Conductivity [W/m-K]
    920,              !- Pipe Density [kg/m3]
    2200,             !- Pipe Specific Heat [J/kg-K]
    0.02667,          !- Pipe Outside Diameter [m]
    0.002413,         !- Pipe Thickness [m]
    Vertical,         !- Heat Exchanger Configuration (Vertical, Horizontal)
    1,                !- Coil Diameter [m]
    0.2,              !- Coil Pitch [m]
    2.5,              !- Trench Depth [m]
    40,               !- Trench Length [m]
    15,               !- Number of Parallel Trenches
    2,                !- Trench Spacing [m]
    15.5,             !- Kusuda-Achenbach Average Surface Temp [C]
    3.2,              !- Kusuda-Achenbach Average Surface Temp Amplitude [C]
    8,                !- Kusuda-Achenbach Phase Shift [C]
    10;               !- Maximum length of simulation [years]
```

to:

```
  GroundHeatExchanger:Slinky,
    Slinky GHX,              !- Name
    GHE Inlet Node,          !- Inlet Node Name
    GHE Outlet Node,         !- Outlet Node Name
    0.0033,                  !- Design Flow Rate {m3/s}
    1.2,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    850,                     !- Soil Specific Heat {J/kg-K}
    1.8,                     !- Pipe Thermal Conductivity {W/m-K}
    920,                     !- Pipe Density {kg/m3}
    2200,                    !- Pipe Specific Heat {J/kg-K}
    0.02667,                 !- Pipe Outer Diameter {m}
    0.002413,                !- Pipe Thickness {m}
    Vertical,                !- Heat Exchanger Configuration
    1,                       !- Coil Diameter {m}
    0.2,                     !- Coil Pitch {m}
    2.5,                     !- Trench Depth {m}
    40,                      !- Trench Length {m}
    15,                      !- Number of Trenches
    2,                       !- Horizontal Spacing Between Pipes {m}
    Site:GroundTemperature:Undisturbed:KusudaAchenbach,  !- Type of Undisturbed Ground Temperature Model
    KATemp 1,                !- Name of Undisturbed Ground Temperature Model
    10;                      !- Maximum Length of Simulation {years}

  Site:GroundTemperature:Undisturbed:KusudaAchenbach,
    KATemp 1,                !- Name
    1.2,                     !- Soil Thermal Conductivity {W/m-K}
    3200,                    !- Soil Density {kg/m3}
    850,                     !- Soil Specific Heat {J/kg-K}
    15.5,                    !- Average Soil Surface Temperature {C}
    3.2,                     !- Average Amplitude of Surface Temperature {deltaC}
    8;                       !- Phase Shift of Minimum Surface Temperature {days}
```

If you notice anything, just tag me.  I'll be intermittent on the weekend, but probably able to do some here and there.

Myoldmopar committed:
Transition rules for ground temp modeling work

Myoldmopar committed:
Merge branch '1579840-GroundTemps' of github.com:mmatts/EnergyPlus into 1579840-GroundTemps

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,125 @@ Site:GroundTemperature:Deep,
+       \minimum> 0.0
+  N4,  \field Soil Moisture Content Volume Fraction
+       \type real
+       \type percent
```
\type percent --> \units percent (below also)

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,125 @@ Site:GroundTemperature:Deep,
+       \maximum 100
+       \default 50
+   N6; \field Evapotranspiration Ground Cover Parameter
+       \type real
```
Would \units dimensionless be appropriate here?

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,125 @@ Site:GroundTemperature:Deep,
+Site:GroundTemperature:Undisturbed:FiniteDifference,
+       \memo Undisturbed ground temperature object using a
+       \memo detailed finite difference 1-D model
+  A1,  \field Name
```
\min-fields 7 (defaults are only processed up through the last field in min-fields)  Check the other new objects as well.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,125 @@ Site:GroundTemperature:Deep,
+  N5,  \field Soil Surface Temperature Amplitude 1
+       \required-field
+       \type real
+       \units deltaC
```
\minimum 0 ? and following

mjwitte replied:
And the phase shift fields probably have a \maximum< 365?  Above objects as well.

mitchute replied:
@mjwitte This one can have amplitudes and phase shifts < 0. I have added it for the above model, though.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,125 @@ Site:GroundTemperature:Deep,
+       \type real
+       \units J/kg-K
+       \minimum> 0.0
+  N4,  \field Average Soil Surface Temperature
```
If these last three fields aren't required and don't have defaults, what happens if left blank?

mitchute replied:
If they are left blank, the model can estimate the parameters from `Site:GroundTemperature:Shallow` which obviously must be included. I added the following note:

```
       \note If left blank the Site:GroundTemperature:Shallow object must be included in the input
       \note The soil temperature, amplitude, and phase shift must all be included or omitted together
```

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1779,26 +1898,15 @@ Site:GroundDomain:Slab,
-        \note If left blank, Site:GroundTemperature:Shallow must be used to provied
-        \note monthly average ground surface temperatures.
-   N12, \field Evapotranspiration Ground Cover Parameter
+   A2,  \field Type of Undisturbed Ground Temperature Model
```
Common pattern it to put "Type" and "Name" and the end of the field name.  e.g. Undisturbed Ground Temperature Model Type and Undisturbed Ground Temperature Model Name (symmetrical fields and no formal type names in the field name).

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1779,26 +1898,15 @@ Site:GroundDomain:Slab,
+   A2,  \field Type of Undisturbed Ground Temperature Model
+        \required-field
+        \type alpha
+        \reference Site:GroundTemperature:Undisturbed:*
```
Delete this \reference

mjwitte replied:
This should be a choice field with a list of valid keys being the three Site:GroundTemperature:Undisturbed . . . . object types.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1779,26 +1898,15 @@ Site:GroundDomain:Slab,
+   A3,  \field Name of Undisturbed Ground Temperature Model
+        \required-field
+        \type alpha
+        \reference Name of Site:GroundTemperature:Undisturbed:*
```
And then this field is an object-list

mjwitte replied:
Lather, rinse, and repeat above comments for other objects.

Commit:
Updating idd fields.

Commit:
Merge branch '1579840-GroundTemps' of github.com:mmatts/EnergyPlus into 1579840-GroundTemps

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -1734,6 +1734,140 @@ Site:GroundTemperature:Deep,
+       \memo detailed finite difference 1-D model
+       \min-fields 7
+  A1,  \field Name
+       \required-field
```
@mmatts Need a \reference UndisturbedGroundTempModels (or similar) on the Name field in all three of the new objects.
 Then all of the "Undisturbed Ground Temperature Model Name" fields need \object-list UndisturbedGroundTempModels

Commit:
Adding \reference and \object-list where needed

Commit:
Merge branch 'develop' into 1579840-GroundTemps

Conflicts:
	doc/src/docs/InputOutputReference/01b-InputOutputReference.md
	tst/EnergyPlus/unit/CMakeLists.txt
	tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc

Commit:
Updating weather manager clear_state()

mjwitte commented:
@Myoldmopar Looks like this introduced some mixed line endings in doc file(s).  The link infiltration branch marked 01b-InputOutputReference as a conflict.  So, I pulled down develop and forced line endings to dos on this file and found a few ground heat transfer lines that showed as diffs.  You have a script to scan all the doc files to clean up any others?

Myoldmopar commented:
Hmmmmmm I thought I had all those fixed. Adding to my list to do in a few minutes.
